src/rust/ide/ast/impl/src/known.rs

```diff
@@ -58,23 +46,48 @@ impl<T> KnownAst<T> {
 
     /// Returns a reference to the stored `Ast` with `Shape` of `T`.
     pub fn ast(&self) -> &Ast { &self.ast }
+}
+
+impl<T,E> KnownAst<T>
+where for<'t> &'t Shape<Ast> : TryInto<&'t T,Error=E> {
+
+    /// Checks if the shape of given Ast node is compatible with `T`.
+    /// If yes, returns Ok with Ast node wrapped as KnownAst.
+    /// Otherwise, returns an error.
+    pub fn try_new(ast:Ast) -> Result<KnownAst<T>,E> {
+        if let Some(error_matching) = ast.shape().try_into().err() {
+            Err(error_matching)
+        } else {
+            Ok(KnownAst {ast,phantom:default()})
+        }
+    }
 
     /// Returns the AST's shape.
-    pub fn shape<E>(&self) -> &T
-    where for<'t> &'t Shape<Ast> : TryInto<&'t T,Error=E>,
-          E                      : Debug, {
+    pub fn shape(&self) -> &T
+    where E : Debug {
```

Comment:
Why this constraint is needed for deref? :O

Reply:
I don't remember. if i checked it or just rewrite it from previous version

Reply:
Without constraint I get:
```error[E0308]: mismatched types
  --> ide/ast/impl/src/known.rs:67:9
   |
67 |         self.deref()
   |         ^^^^^^^^^^^^ expected type parameter, found struct `known::KnownAst`
   |
   = note: expected type `&T`
              found type `&known::KnownAst<T>`
   = help: type parameters must be constrained to match other types
   = note: for more information, visit https://doc.rust-lang.org/book/ch10-02-traits.html#traits-as-parameters```
This is really strange to me